feat(parametermanager): Added samples for create global parameter and parameter version

durgesh-ninave-crest opened the pull request:
## Description

Created samples for creating parameter and parameter version using Parameter Manager SDK

#### Sample List:

1. [parameter-structured-data](https://cloud.google.com/secret-manager/parameter-manager/docs/create-parameter#parameter-structured-data)
2. [parameter-unstructured-data](https://cloud.google.com/secret-manager/parameter-manager/docs/create-parameter#parameter-unstructured-data)
3. [add-parameter-version](https://cloud.google.com/secret-manager/parameter-manager/docs/add-parameter-version#add-parameter-version) (structured, unstructured)
4. [create-parameter-secret-reference](https://cloud.google.com/secret-manager/parameter-manager/docs/reference-secrets-in-parameter#create-parameter-secret-reference)

amanda-tarafa reviewed (changes requested):
All the other samples are in the parametermanager/api folder. That's where these should be as well.

Also please address here my latest coments on #2929 regarding the fixture being used for shared data only. Data that's on the fixture it to be shared across tests that do not modify it. This includes the resource names/ids.

durgesh-ninave-crest commented:
> All the other samples are in the parametermanager/api folder. That's where these should be as well.
> 
> Also please address here my latest coments on #2929 regarding the fixture being used for shared data only. Data that's on the fixture it to be shared across tests that do not modify it. This includes the resource names/ids.

Moved the samples and tests to `parametermanager/api`, and cleaned up the test cases so the fixture just holds shared stuff like the create, delete methods.

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateParameterTests.cs:
```diff
@@ -0,0 +1,44 @@
+        Assert.Equal(result.ParameterName.ParameterId, parameterName.ParameterId);
+        Assert.Equal(ParameterFormat.Unformatted, result.Format);
+
+        _fixture.ParametersToDelete.Add(parameterName);
```
See my note on the fixture. It's better to add the parameter on the list as soon as you create it, so that i gets deleted even if some of the Asserts fail.

durgesh-ninave-crest replied:
Good point — I’ve updated the code to add the parameter and version to the list right after creation, so cleanup still happens even if an assert fails.

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateParameterVersionTests.cs:
```diff
@@ -0,0 +1,46 @@
+        ParameterVersionName parameterVersionName = new ParameterVersionName(_fixture.ProjectId, ParameterManagerFixture.LocationId, _fixture.RandomId(), _fixture.RandomId());
+        Parameter parameter = _fixture.CreateParameter(parameterVersionName.ParameterId, ParameterFormat.Unformatted);
+        string payload = "test123";
+        ParameterVersion result = _sample.CreateParameterVersion(
+          projectId: parameterVersionName.ProjectId, parameterId: parameterVersionName.ParameterId, versionId: parameterVersionName.ParameterVersionId, payload: payload);
```
(Indent 4 spaces always)

```suggestion
        ParameterVersion result = _sample.CreateParameterVersion(
            projectId: parameterVersionName.ProjectId, parameterId: parameterVersionName.ParameterId, versionId: parameterVersionName.ParameterVersionId, payload: payload);
```

durgesh-ninave-crest replied:
Fixed missing spaces in the code.

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateParameterVersionTests.cs:
```diff
@@ -0,0 +1,46 @@
+        Assert.Equal(result.ParameterVersionName.ParameterVersionId, parameterVersionName.ParameterVersionId);
+
+        _fixture.ParametersToDelete.Add(parameter.ParameterName);
+        _fixture.ParameterVersionsToDelete.Add(parameterVersionName);
```
Since we don't create the version in the fixture (as that's what we are testing), do this right after creating the version, so that if any of the asserts fail we still clean up.

durgesh-ninave-crest replied:
Updated the code to add the parameter and version to the list right after creation, so cleanup still happens even if an assert fails.

amanda-tarafa commented:
Also, most of the change requests from my latest review apply to several of the files on this PR (and presumably to the other PRs as well), even if they are just called out in a single file.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+
+    public String RandomId()
```
Changed the return type from `String` to `string`.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,95 @@
+    public const string LocationId = "global";
+
+    internal List<ParameterName> ParametersToDelete = new List<ParameterName>();
+    internal List<ParameterVersionName> ParameterVersionsToDelete = new List<ParameterVersionName>();
```
I've Added `{ get; }` to both lists.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,95 @@
+
+    public Parameter CreateParameter(string parameterId, ParameterFormat format)
+    {
+        ParameterManagerClient client = ParameterManagerClient.Create();
```
I've initialised the client in the constructor and removed its creation from the other methods.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,95 @@
+            Format = format
+        };
+
+        return client.CreateParameter(projectName, parameter, parameterId);
```
I've added the parameter to the list for deletion.

amanda-tarafa commented:
@durgesh-ninave-crest Mention me in a comment once this is ready for re-review. Thanks!

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples/CreateStructuredParameterVersion.cs:
```diff
@@ -0,0 +1,67 @@
+    /// <param name="versionId">The ID of the version to be created.</param>
+    /// <param name="payload">The JSON dictionary payload to be stored in the new parameter version.</param>
+    /// <returns>The created ParameterVersion object.</returns>
+    public ParameterVersion CreateStructuredParameterVersion(
```
Thank you for pointing that out. Since the code in both samples was identical, I've revised the samples to include unformatted and JSON payload within the samples.

durgesh-ninave-crest commented:
> @durgesh-ninave-crest Mention me in a comment once this is ready for re-review. Thanks!

@amanda-tarafa This PR is ready for re-review. Once the other PRs are ready, I will ping you.

durgesh-ninave-crest commented:
> @amanda-tarafa This PR is ready for re-review. Once the other PRs are ready, I will ping you.

All the PRs are ready for review. As you mentioned in https://github.com/GoogleCloudPlatform/dotnet-docs-samples/pull/2930#issuecomment-2797346069, 2 PRs are open, and the others have been converted to drafts.

amanda-tarafa commented:
Perfect, thanks! I'll review again sometime this week.

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateStructuredParameterVersionTests.cs:
```diff
@@ -0,0 +1,44 @@
+        _fixture.ParameterVersionsToDelete.Add(parameterVersionName);
+
+        Assert.NotNull(result);
+        Assert.Equal(result.ParameterVersionName.ParameterVersionId, parameterVersionName.ParameterVersionId);
```
I think the order of the parameters is swaped, the first parameter is the expected value and the second parameter is the actual value.
```suggestion
        Assert.Equal(parameterVersionName.ParameterVersionId, result.ParameterVersionName.ParameterVersionId);
```

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateStructuredParameterTests.cs:
```diff
@@ -0,0 +1,43 @@
+
+        Assert.NotNull(result);
+        Assert.Equal(ParameterFormat.Json, result.Format);
+        Assert.Equal(result.ParameterName.ParameterId, parameterName.ParameterId);
```
Swaped parameters in Assert.

This is important because of how test failures are reported. Please review/modify everywhere.

durgesh-ninave-crest replied:
Updated the assertion

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateParameterVersionWithSecretTests.cs:
```diff
@@ -0,0 +1,44 @@
+        _fixture.ParameterVersionsToDelete.Add(parameterVersionName);
+
+        Assert.NotNull(result);
+        Assert.Equal(result.ParameterVersionName.ParameterVersionId, parameterVersionName.ParameterVersionId);
```
Swap these

durgesh-ninave-crest replied:
Updated the assertion

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateParameterVersionTests.cs:
```diff
@@ -0,0 +1,43 @@
+        _fixture.ParameterVersionsToDelete.Add(parameterVersionName);
+
+        Assert.NotNull(result);
+        Assert.Equal(result.ParameterVersionName.ParameterVersionId, parameterVersionName.ParameterVersionId);
```
Swap

durgesh-ninave-crest replied:
Updated the assertion

amanda-tarafa commented on parametermanager/api/ParameterManager.Samples.Tests/CreateParameterTests.cs:
```diff
@@ -0,0 +1,43 @@
+        _fixture.ParametersToDelete.Add(parameterName);
+
+        Assert.NotNull(result);
+        Assert.Equal(result.ParameterName.ParameterId, parameterName.ParameterId);
```
Swap

durgesh-ninave-crest replied:
Updated the assertion

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,97 @@
+    public string ProjectId { get; }
+    public const string LocationId = "global";
+
+    public ParameterManagerClient client { get; }
```
Changed `client` to `Client`.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,97 @@
+            throw new Exception("missing GOOGLE_PROJECT_ID");
+        }
+
+        client = ParameterManagerClient.Create();
```
Changed `client` to `Client`.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,97 @@
+            Format = format
+        };
+
+        Parameter Parameter = client.CreateParameter(projectName, parameter, parameterId);
```
Changed `client` to `Client`.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,97 @@
+    {
+        try
+        {
+            client.DeleteParameter(name);
```
Changed `client` to `Client`.

durgesh-ninave-crest commented on parametermanager/api/ParameterManager.Samples.Tests/ParameterManagerFixture.cs:
```diff
@@ -0,0 +1,97 @@
+    {
+        try
+        {
+            client.DeleteParameterVersion(name);
```
Changed `client` to `Client`.

amanda-tarafa reviewed (approved):
These look good, thanks again for the extra changes.

I'll merge once tests are green.
